### ISSUE TITLE
Will this be Open Source?

### ISSUE THREAD

**Issue by sirinath** [microsoft/WSL#1]
Hopefully Bash On Windows will be opensource

**Comment by bitcrazed** [comment]
We have no firm plans as yet, but we're not averse to open-sourcing some of this tech.

**Comment by sirinath** [comment]
Hopefully you will be slanted towards open sourcing it with well documented internals to get other subsystems like Darwin, BSD, SmartOS or other systems running on Windows.

**Comment by BSalita** [comment]
There's lots of issues with terminal emulation in the first beta. This is particularly an area where open source availability would quicken the pace of bug fixing and compatibility.

**Comment by poizan42** [comment]
Can I vote for having lxcore.sys open? Just think of the possibilities in having a starting point for the community to implement other subsystems.

(Am I wrong in assuming that implementing a subsystem requires tight coupling with kernel internals that you don't want to have to keep that way for backwards compatibility in the future?)

**Comment by russalex** [comment]
Great conversation around open sourcing and we hear the requests.  Might I suggest voting on this on our [uservoice page](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo)?

**Comment by raeesiqbal** [comment]
There is actually a very similar open-source [project called flinux](https://github.com/wishstudio/flinux). Whic is based on the same idea of translating Linux Syscalls to Windows NT Syscalls. Mabye Windows team can get some inspiration from them :smile:

**Comment by raeesiqbal** [comment]
I've just created a [UserVoice feedback for this idea](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13389477-make-it-windows-subsystem-for-linux-open-source). Anyone interested can vote for this there.

**Comment by ytrezq** [comment]
@raeesbhatti : it convert Linux calls directly to ɴᴛ calls. This is different since the windows subsystem *(`win32k.sys`)*is completely bypassed in favor of`lxcore.sys` *(yes, windows is subsystem itself)*.

**Comment by ssokolow** [comment]
@ytrezq: Yeah. A lot of people don't know that Windows NT was designed to be modular that way. It was originally designed to have three subsystems: Win32, POSIX, and OS/2.

Heck, drive letters are an artifact of the Win32 subsystem and NTFS can operate in a POSIX mode where the only characters disallowed in file names are NUL and /

**Comment by ytrezq** [comment]
@ssokolow : as I’m currently looking for a way to compile a native ᴅʟʟ for the ᴘᴘᴄ architecture without original hardware, I perfectly know that :-)

**Comment by ssokolow** [comment]
I suspected as much. I probably should have been a bit more clear that my comment was for other people's benefit.

**Comment by slavanap** [comment]
@bitcrazed What about disclosing source code of `/init` binary? It uses it *must* use some of Ubuntu/Linux sources because it depends on `/lib64/ld-linux-x86-64.so`

**Comment by ssokolow** [comment]
@slavanap 
All dynamically-linked Linux binaries depend on some version of `libc.so` and glibc's license specifically contains an exception to allow that for closed-source applications.

**Comment by russalex** [comment]
Long story short, we are still working on /init.  What is out there was required to get WSL out the door, but there are some pretty low level changes we need to make for future features.  We're just not in a position to discuss open sourcing it yet.

We have had the larger open source conversation a number of times in house but the same story goes there.  We have some new features that still need to be made before we can truly discuss feasibility and what parts of the project we can / make sense to open source.  None of us are opposed to doing it, we just need to do it when it makes sense and do it right.

**Comment by erkinalp** [comment]
> Long story short, we are still working on /init. 

Systemd2svchost.

**Comment by erkinalp** [comment]
You will see if Windows is opensourced by looking up the status code of https://github.com/Microsoft/Windows
If it returns 404, no. If it returns 503, it is and GitHub is overcrowded. If it returns 200, it is and GitHub has been able to cope with crowd.

**Comment by DarthSpock** [comment]
@erkinalp There's a difference between open-sourcing WSL and open-sourcing Windows 10. I do not expect them to open-source the OS (though that would be nice too, a fantasy dream though), just the Linux subsystem portion.

**Comment by erkinalp** [comment]
> There's a difference between open-sourcing WSL and open-sourcing Windows 10
It is impossible to opensource WSL but not Windows 10 kernel under a share-alike free software license.

**Comment by ssokolow** [comment]
@erkinalp There are two easy ways to open-source WSL without open-sourcing Windows 10.

1. License it under "GPL plus a [linking exception](https://en.wikipedia.org/wiki/GPL_linking_exception)".

    Normally, you say "I grant you permission to use this under the terms of the GPL". A linking exception lets you say something like "I grant you permission to use GNU libc under the terms of the GPL, but I also grant you permission to compile proprietary programs against it without forcing the proprietary code to be open source." 

    A linking exception could say "but I also grant you permission to link it against Microsoft's proprietary code without requiring that code to be released under the GPL."

    Then, the generally understood "If you contribute back, you agree to license your contributed code to the project under the same terms as you received" terms of licensing would Just Work™ as "GPL... but it can also be linked into Windows".

2. Require copyright assignment

    If everything is owned by Microsoft, they can do whatever they want with it. (You're never bound by the license you offer to others beyond clauses in the GPL like "non-revokable". It's the "contributed patches are licensed TO you by others" that locks a license in.)

    If contributors sign over the rights to their code, then it's legally no different than Microsoft writing it all and there's no impediment to Microsoft "GPLing one copy and compiling another into the kernel".

    That's why it's possible for a sole rightsholder, like Sun Microsystems, to release a project like OpenOffice.org under a copyleft license, yet also sell proprietary licenses to IBM.

**Comment by DarthSpock** [comment]
To back what @ssokolow has said, Microsoft HAS open-sourced other applications such as Powershell and VSCode. Powershell specifically is deeply embedded in later versions of Windows because it's the replacement of CMD and they've only left CMD in place for those who might favor it or have more experience using it. MS made it clear though that you have to know Powershell via certifications. But now you can install on any system with the only limitation being how fast they can make it available to everyone while fixing all the bugs that have resulted from making it open source. WSL will be slightly different in this case because of the underlying infrastructure is supposed to be based on a system that's already publicly known how to work for decades. This is why we have so many distros available and more can continue to be developed. All one needs to do is compare the code used to make something POSIX compliant and make something map to it inside the NT kernel. The WSL team themselves have said they try to use Windows APIs for everything to make things easier and of course provides better interop. Anybody can look at those APIs because without them, there would be no software developed for Windows. We're not asking for a huge jump here, it's simply a natural extension of everything MS has already done. If they want to take it one step further and make the whole OS open-source great! Eventually MS might find itself in that rabbit hole but as long as they're still able to make profit off of Windows, I doubt they'll open-source it.

**Comment by erkinalp** [comment]
>    License it under "GPL plus a linking exception".

Impossible as all Windows team can see all Windows source code. They are not separate.

**Comment by bitcrazed** [comment]
@DarthSpock @ssokolow We appreciate the enthusiasm here, but there's A LOT of background that currently prevents us from open-sourcing WSL. 

Note that WSL is almost all NT kernel code which calls and takes advantage of as many NT Kernel subsystems, structures, behaviors, etc., as possible, implementing a few Linux specific behaviors/data where necessary. The APIs WSL takes advantage of are NT Kernel APIs, not the Win32 APIs used by most Windows apps. 

Also, WSL takes advantage of NT Kernel protection mechanisms that prevent the loading of binaries that have not been signed with the appropriate keys. This would prevent any non-MSFT developer from even executing their modified WSL binaries, let alone testing them. Therefore, developers would have to submit non-tested PR's which, of course, are unlikely to be accepted.

Again, appreciate the passionate debate here, but understand that open-sourcing a key kernel subsystem that is deeply integrated with many other parts of the NT kernel is no easy undertaking, even if the non-technical issues could be solved.

**Comment by poizan42** [comment]
@bitcrazed You can just bypass the signature check if you have a kernel debugger attached, so that point seems a bit moot. Obviously it prevents people from running third-party versions normally, but people can still test their modifications out. I understand the other issues making it difficult to open-source.

**Comment by erkinalp** [comment]
> Note that WSL is almost all NT kernel code which calls and takes advantage of as many NT Kernel subsystems, structures, behaviors, etc., as possible, implementing a few Linux specific behaviors/data where necessary. The APIs WSL takes advantage of are NT Kernel APIs, not the Win32 APIs used by most Windows apps.

Why not open source whole Windows 10, including WSL?

**Comment by Biswa96** [comment]
@benhillis Does MS programmers use any GNU code in `LxCore.sys` or `init`? When I type `cat /proc/version` WSL shows a string with `gcc version 5.4.0 (GCC)`. LxCore.sys has this function `LxpProcFsRootVersionGenerateBuffer()`.

**Comment by erkinalp** [comment]
@HBelusca: Why are you confused? Release of Windows source code under GPL2 will benefit both Windows and ReactOS.
p.s.: Never worked at MS, never seen Windows source code, a proud member of ReactOS community

**Comment by benhillis** [comment]
@biswa96 - no GNU (I think you mean GPL?) code is used in our driver or init daemon. The kernel version string is a lie, but we should probably remove the gcc bit since it's produced by Microsoft's compilers and not gcc. It's there because early in the project we thought usermode binaries may parse that string and we wanted to be similar to what Ubuntu's kernel reported.

**Comment by erkinalp** [comment]
@bitcrazed Is there an update on this?
@HBelusca opensource of Windows it will be good for ReactOS community. We would be Windows developers and ReactOS developers at the same time.

**Comment by erkinalp** [comment]
@bitcrazed Is there an update on this?

**Comment by benhillis** [comment]
@erkinalp - This question is WAY above our paygrade so we cannot speculate.

**Comment by erkinalp** [comment]
Windows and ReactOS are currently separate projects which cannot look at each other's source code. When Windows is open sourced under GPL 2, they would be able to cooperate like Debian and Ubuntu do (though, they are different in one regard; Ubuntu is a derivative of Debian, ReactOS is reimplementation of Windows). Suddenly, 300k more Windows developers for free :D

**Comment by bitcrazed** [comment]
@erkinalp Appreciate the enthusiasm here, but no, we have no plans at present to opensource WSL or Windows.

If we ever change this position, I am pretty sure you'll hear about it via the monumental amount of news that such an announcement would generate - no need to revisit this thread to find out!

**Comment by leon0399** [comment]
I love how this issue is accidentally referenced in hundreds of other issues 😅

**Comment by Kreijstal** [comment]
narrator: it was not open source

**Comment by misaka00251** [comment]
Yes finally
https://github.com/microsoft/WSL/commit/697572d664c9371abbc53cab9cf7b4b32be1980e

**Comment by benhillis** [comment]
After nine years, I am thrilled to announce that Windows Subsystem for Linux is now open source!

I would like to take this opportunity to extend my heartfelt gratitude to our enthusiastic community of users and Linux distro partners. Your feedback and contributions have been instrumental in shaping WSL into what it is today.

**Comment by salmanmkc** [comment]
yay!

**Comment by AlekSi** [comment]
@benhillis, you may want to update the repo description:

![Image](https://github.com/user-attachments/assets/2c6b24fe-79b4-4822-813c-52363e548b39)

Congrats!

**Comment by benhillis** [comment]
> [@benhillis](https://github.com/benhillis), you may want to update the repo description:
> 
> ![Image](https://github.com/user-attachments/assets/2c6b24fe-79b4-4822-813c-52363e548b39)
> 
> Congrats!

Good call! Lots of stuff to update today...

**Comment by redyf** [comment]
LETS GO! Open Source for the win

**Comment by AriESQ** [comment]
We did it team! 🎉🎊🎉

**Comment by Cattn** [comment]
W

**Comment by brasky** [comment]
W

**Comment by KristijanZic** [comment]
Very nice :)

**Comment by Biswa96** [comment]
Are every source file added to this repository? For example, symbols from src\windows\common\lxssbusclient.cpp are missing.

**Comment by vicradon** [comment]
Hehe, historical

**Comment by lsevero** [comment]
🚀🚀🚀

**Comment by Yovach** [comment]
let's go

**Comment by fossdd** [comment]
It's about time. Leveriging open source behind closed-source-software has been a terrible decision to begin with.

**Comment by aspizu** [comment]
is this going to be deprecated?

**Comment by yhoungdev** [comment]
Finally, it's open-source, and now this issue should be like a digital art, framed to the wall of microsoft HQ.

**Comment by darwingtx** [comment]
🤯🎉

**Comment by LuminescentBulb** [comment]
Chat LET'S GOOOOO

**Comment by Tina-otoge** [comment]
Rare Windows W, congratulations to the team and all parties involved in the open-sourcing of WSL components, this is huge!

**Comment by Kreijstal** [comment]
press W in the chat for the rizzlers at microsoft who open source the wsl. no cap

**Comment by yhoungdev** [comment]
> > Rare Windows W, congratulations to the team and all parties involved in the open-sourcing of WSL components, this is huge!
> 
> There's probably a catch. I wouldn't get my hopes up.

Catches like ?

**Comment by Kreijstal** [comment]
> > > Rare Windows W, congratulations to the team and all parties involved in the open-sourcing of WSL components, this is huge!
> > 
> > 
> > There's probably a catch. I wouldn't get my hopes up.
> 
> Catches like ?

the source code of Lxcore.sys

**Comment by jklgrasso** [comment]
w

**Comment by kinduff** [comment]
I was here! Great move 👏

**Comment by xnox** [comment]
It has code from partners too? cause i thought i contributed to that code too whilst at Canonical, for the initial launches.

**Comment by zadorski** [comment]
Kudos to the WSL team! What's next? :)
Upstream WSL paravirtual interfaces into the Linux kernel?

**Comment by hasanalikhattak** [comment]
Amazing times we are seeing.

**Comment by trungnt2910** [comment]
9 years ago the original issue was about WSL1, or the original stack that relies on `lxcore.sys`.

Today, with `lxcore.sys` still not included, it's not entirely solved yet: #12922.

**Comment by elvessilvavieira** [comment]
DONE

**Comment by Skywalker8510** [comment]
Heck yea. A great move by Microsoft.

**Comment by arefdsg** [comment]
oo

**Comment by iwangbowen** [comment]
Yes

**Comment by sanjaysah101** [comment]
Excited to see code base and contribute to it. 🤩

**Comment by Anutrix** [comment]
Thank You.

**Comment by SuperKenVery** [comment]
Hooray! Better than ever!

**Comment by hohoaisan** [comment]
Let's goo!

**Comment by stong** [comment]
W microsoft

**Comment by DavidXanatos** [comment]
why not open source:
    Lxcore.sys, the kernel side driver that powers WSL 1
and
    P9rdr.sys and p9np.dll, which runs the “\\wsl.localhost” filesystem redirection (from Windows to Linux)

Please opensource these as well.

**Comment by ViktorPopp** [comment]
Nice!

**Comment by PeerFixer** [comment]
This is truly incredible🥰

**Comment by cdevroe** [comment]
👏 bravo

**Comment by looshch** [comment]
\*9 years later\*

**Comment by IAmTheNerdNextDoor** [comment]
> *9 years later*

Microsoft took a page out of Valve's book... "In Microsoft time" is real guys.

**Comment by younes-io** [comment]
This thread should be in history books

**Comment by soulis-1256** [comment]
Great job to the people that made this happen!

**Comment by FayeSpica** [comment]
Historical moments👏

**Comment by qdx** [comment]
It always amazes me to see such long term efforts flourish, congratulations for thoes who pushed along the way, I admire you and envy you at the same time.

**Comment by msundin** [comment]
Great news! 🙌🏻

**Comment by raulcarlomagno** [comment]
Aleluya! 🎉

**Comment by nekiwo** [comment]
just wanted to leave a reply at this historic moment🫡

**Comment by khatrivarun** [comment]
hell yeah !!

**Comment by NaveenPNair** [comment]
I was here. 🙌

**Comment by forkthat** [comment]
Such a historic moment! Thank you, guys! 🙌🥂

**Comment by sanamhub** [comment]
COOL

**Comment by yashasjindal** [comment]
Rare microsoft W

**Comment by t2hv33** [comment]
Cool moment

**Comment by Certezalito** [comment]
Stay cool everyone and please sign the yearbook! : https://github.com/microsoft/WSL/issues/1

**Comment by marbaa** [comment]
I don't understand why people are so happy that it is open sourced. We can assume that WSL will go to RIP little bit faster now than when it was closesourced. Basically we can see that it is already, only people in issues discussions are people with problems and then bots. For years. Nobody from devs.

There are still big problems with WSL networking general, with networking after sleep/hibernate, with VPNs after so many years and with all respect to community developers, I doubt that they will be able to fix them, when MS developers weren't able/didn't cared.

edit: I would like to see arguments from downvoters.

**Comment by fzazaza** [comment]
```

```

**Comment by fzazaza** [comment]
- [ ]

**Comment by sanamhub** [comment]
> Hopefully Bash On Windows will be opensource

<img width="631" height="633" alt="Image" src="https://github.com/user-attachments/assets/c4a63fdf-8fbe-4661-a59b-eca0eaeb587b" />

**Comment by bdbdnrnrb-afk** [comment]
ico
